metadata-ingestion/src/datahub/api/entities/external/external_entities.py

```diff
@@ -0,0 +1,252 @@
+            if add_to_cache:
+                self.cache[platform_resource.id] = platform_resource
+            yield platform_resource
+
+    def create(self, platform_resource: PlatformResource) -> None:
+        platform_resource.to_datahub(self.graph)
+        self.cache[platform_resource.id] = platform_resource
+
+    def get(self, key: PlatformResourceKey) -> Optional[PlatformResource]:
+        return self.cache.get(key.id)
+
+    def delete(self, key: PlatformResourceKey) -> None:
+        self.graph.delete_entity(urn=PlatformResourceUrn(key.id).urn(), hard=True)
+        del self.cache[key.id]
+
+
+class ExternalEntityId:
+    """
+    ExternalEntityId is a unique
+    identifier for an ExternalEntity.
+    """
+
+    @abstractmethod
+    def to_platform_resource_key(self) -> PlatformResourceKey:
+        """
+        Converts the ExternalEntityId to a PlatformResourceKey.
+        """
+        pass
+
+
+class CaseSensitivity(Enum):
+    UPPER = "upper"
+    LOWER = "lower"
+    MIXED = "mixed"
+
+
+class LinkedResourceSet(BaseModel):
+    """
+    A LinkedResourceSet is a set of DataHub URNs that are linked to an ExternalEntity.
+    """
+
+    urns: List[str]
+
+    class Config:
+        arbitrary_types_allowed = True
+
+    @staticmethod
+    def detect_case_sensitivity(urn: Urn) -> CaseSensitivity:
+        if urn.get_entity_id_as_string().isupper():
+            return CaseSensitivity.UPPER
+        elif urn.get_entity_id_as_string().islower():
+            return CaseSensitivity.LOWER
+        return CaseSensitivity.MIXED
+
+    def _detect_case_sensitivity(self) -> Optional[CaseSensitivity]:
+        """
+        Detects the case sensitivity of the URNs in the set.
+        """
+        if len(self.urns) == 0:
+            return CaseSensitivity.MIXED
+
+        if all(
+            LinkedResourceSet.detect_case_sensitivity(Urn.from_string(urn))
+            == CaseSensitivity.UPPER
+            for urn in self.urns
+        ):
+            return CaseSensitivity.UPPER
+        elif all(
+            LinkedResourceSet.detect_case_sensitivity(Urn.from_string(urn))
+            == CaseSensitivity.LOWER
+            for urn in self.urns
+        ):
+            return CaseSensitivity.LOWER
+        return CaseSensitivity.MIXED
+
+    def _has_conflict(self, urn: Urn) -> bool:
+        """
+        Detects if the urn is safe to add into the set
+        This is used to detect conflicts between DataHub URNs that are linked to
+        the same ExternalEntity.
+        e.g. Case sensitivity of URNs
+        Mixing tags and terms in the same set etc.
+        Return True if the urn is not safe to add into the set, else False.
+        If the urn is already in the set, we don't need to add it again, but
+        that is not a conflict.
+        """
+        if urn.urn() in self.urns:
+            return False
+
+        # Detect the family of the urns in the existing set
+        detected_family = None
```

Comment:
`family` refers here to `entity_type`, right? if so, why not making use of `entity_type` terminology instead?

Also, is it correct my understanding that a `LinkedResourceSet` is restricted to a single family, _ie_ entity type? if so, why not making that explicit in the constructor of the `LinkedResourceSet`?

```
class LinkedResourceSet(BaseModel):
    """
    A LinkedResourceSet is a set of DataHub URNs that are linked to an ExternalEntity.
    """

     entity_type: str
    urns: List[str]
```

That would remove some complexity in this `_has_confict` method?